Check added stack trace parts for filename match

fongandrew opened the pull request:
Fixes https://github.com/substack/tape/issues/386. Also refactors regex checking code a little.

One thing I'm not sure about is how to test this. The issue is actually this line here:
https://github.com/fongandrew/tape/blob/2c242ca9715b6b0bfded23fccb522dd414d96612/lib/test.js#L219

Tape ignores stack trace lines that originate from files starting with that `dir` variable. Currently, it means ignoring all stack trace lines from within tape's own package directory. The problem is that tape's own tests also live within its package directory, so the `at` variable is never set on the result object.

We could try changing `dir` to point to just the `lib` directory inside the package. This would result in the `at` being set and we could check it. This should be fine to do, since the stack trace for an assertion error shouldn't run through any file outside of the `lib` directory, but I'm not sure if that could change in the future.

Also, this would result in 25 or so existing tests failing (since their test outputs don't include the `at` variable). It wouldn't be a big deal to update those tests (correct behavior should be including the `at` variable in those test results), but feedback welcome on that.

ljharb commented:
I think that tests kind of need to *not* ignore stack trace lines that originate within the tests directory, so I'd think updating those tests is the right path here (also totally fine to do that in a separate PR, and this can be rebased after that's merged)

fongandrew commented:
Updated tests are in https://github.com/substack/tape/pull/388. This PR is rebased off of that one.

ljharb commented:
Please rebase this on latest master when you can :-D

fongandrew commented:
Rebased!

ljharb commented on lib/test.js:
```diff
@@ -225,29 +225,25 @@ Test.prototype._assert = function assert (ok, opts) {
+            var filemRe = /((?:\/|[A-Z]:\\)[^:\s]+:(\d+)(?::(\d+))?)/;
+            var filem;
+            var sIndex;
+            for (sIndex in s.slice(0, 4)) {
```
for..in loops should be avoided; this would be better done as a for loop, or as `Array.prototype.some.call` or something

fongandrew replied:
`Array.prototype.forEach` would be the cleanest way to do this, but it's not supported in older versions of Node (or older browsers) without polyfill support. Not sure what Tape's polyfill policy is.

I can switch it to some variation of `for (sIndex = 0; sIndex < s.slice(0, 4).length; s++) { ... }` though.

(I'm not sure what expectations Tape makes

ljharb replied:
True; we have the `for-each` package to handle `forEach` but that doesn't handle `some`.

ljharb commented on lib/test.js:
```diff
@@ -225,29 +225,25 @@ Test.prototype._assert = function assert (ok, opts) {
             if (filem[3]) res.column = filem[3];
             
-            res.at = m[1];
+            res.at = s.length > 1 ? m[1] : '<anonymous> (' + m[1] + ')';
```
why s.length instead of m.length here?

fongandrew replied:
`m` is the result of the regular expression matching something like `at /path/to/file.js:123:10` or `at functionName /path/to/file.js:123:10`. If we're at this point in the code, `m.length` should always be the same, where `m[1]` is the portion of the regex matching everything after `at`.

`s` is `m[1].split(/\s+/)` So if `s` has a length of 1, we know that`m[1]` is just case where there's just a path with no function name.

Although in writing this, it now occurs to me that the original `m[1].split(/\s+/)` code fails to account for directory and file names with spaces ... lemme go fix that.

ljharb replied:
Maybe we could store `s.length > 1` in a nicely named variable, for clarity?

fongandrew commented:
Added test for a file name with a space in it. Rewrote the stack trace regex to handle that scenario. New approach also gets rid of the `for ... in` loop.

ljharb commented on lib/test.js:
```diff
@@ -219,35 +219,59 @@ Test.prototype._assert = function assert (ok, opts) {
+                at Test.bound [as run] (/path/to/file.js:123:45)
+                at /path/to/file.js:123:45
+            */
+            var m = new RegExp(
```
it would be really nice if we could use a regex literal, and repeat the parts you want to comment, in the comment, instead of using `new RegExp`

ljharb commented on lib/test.js:
```diff
@@ -219,35 +219,59 @@ Test.prototype._assert = function assert (ok, opts) {
+            res.functionName = callDescription.split(/s+/)[0]
+            res.file = filePath;
+            res.line = Number(m[3]);
+            if (m[4]) res.column = m[4];
```
`Number(m[4])`?

fongandrew commented:
Comments addressed (amended old commit).
